### PR TITLE
Run circular-dependencies plugin during build

### LINKED ISSUE #375: investigate webpack + circular imports
**From:** https://github.com/cylc/cylc-ui/pull/359#discussion_r368371454
**See also:** https://railsware.com/blog/how-to-analyze-circular-dependencies-in-es6/

Webpack can have issues making ES6 circular imports synchronous, we should analyse our codebase to ensure we don't hit this issue.

**Pull requests welcome!**

### PR DESCRIPTION
**Pull request by kinow**
These changes close #375 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Does not need tests (why? build only).
- [x] No change log entry required (why? e.g. invisible to users).
- [x] No documentation update required.

### CODE REVIEW

File: vue.config.js [hunk 1]
```diff
@@ -30,8 +31,33 @@ module.exports = {
...
+        allowAsyncCycles: false,
+        cwd: process.cwd(),
+        onDetected ({ module: webpackModuleRecord, paths, compilation }) {
+          compilation.errors.push(new Error(`Cyclic dependency: ${paths.join(' -> ')}`))
+        }
+      })
+    ]
+  },
   chainWebpack: config => {
-    if (process.env.VUE_APP_SERVICES === 'offline' || process.env.NODE_ENV === 'test') {
```
Comment (kinow) [inline vue.config.js:34]: Just organizing code below, feel free to ignore. Important changes are above :point_up:

**Comment by kinow** [comment]
CI should fail. Here's my environment output:

```bash
kinow@ranma:~/Development/python/workspace/cylc-ui$ yarn run build
yarn run v1.22.5
$ vue-cli-service build

⠹  Building for production...

 ERROR  Failed to compile with 19 errors                                                                                                    3:28:10 PM

 error  

Cyclic dependency: src/components/cylc/tree/cylc-tree.js -> src/components/cylc/tree/index.js -> src/components/cylc/tree/cylc-tree.js

 error  

Cyclic dependency: src/components/cylc/tree/index.js -> src/components/cylc/tree/cylc-tree.js -> src/components/cylc/tree/index.js

 error  

Cyclic dependency: src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue

 error  

Cyclic dependency: src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/Enum.vue -> src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/Enum.vue

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/Enum.vue -> src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/Enum.vue -> src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/List.vue -> src/components/graphqlFormGenerator/components/List.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/List.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/List.vue

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/List.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/List.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/List.vue -> src/components/graphqlFormGenerator/components/List.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/List.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/List.vue -> src/components/graphqlFormGenerator/components/List.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/List.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/NonNull.vue -> src/components/graphqlFormGenerator/components/NonNull.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/NonNull.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/NonNull.vue

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/NonNull.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/NonNull.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/NonNull.vue -> src/components/graphqlFormGenerator/components/NonNull.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/NonNull.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/NonNull.vue -> src/components/graphqlFormGenerator/components/NonNull.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/NonNull.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/Object.vue -> src/components/graphqlFormGenerator/components/Object.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/Object.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/Object.vue

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/Object.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/Object.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/Object.vue -> src/components/graphqlFormGenerator/components/Object.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/Object.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/Object.vue -> src/components/graphqlFormGenerator/components/Object.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/Object.vue?vue&type=script&lang=js&

 error  

Cyclic dependency: src/components/graphqlFormGenerator/components/vuetify.js -> src/components/graphqlFormGenerator/components/Enum.vue -> src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/Enum.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/components/vuetify.js

 error  

Cyclic dependency: src/components/graphqlFormGenerator/mixins.js -> src/components/graphqlFormGenerator/FormInput.vue -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/FormInput.vue?vue&type=script&lang=js& -> src/components/graphqlFormGenerator/mixins.js

 ERROR  Build failed with errors.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

```

**Comment by kinow** [comment]
One of the circular dependencies, from the tree component is fixed in #599

**Comment by kinow** [comment]
Ignored the form generator components as they seem to be intentionally tightly coupled. Build passing, and will scan for future circular dependencies in PR's.

I think GH actions are still not being triggered, so we may need to trigger them later when the Azure issue has been fixed.

**Comment by kinow** [comment]
GH actions build OK

**Review by hjoliver — APPROVED** [review]
LGTM :+1:

**Review by oliver-sanders — APPROVED** [review]
Thanks.

**Pull request merged**